raiden/network/protocol.py

```diff
@@ -18,23 +18,343 @@
     UnknownAddress,
     UnknownTokenAddress,
 )
+from raiden.constants import (
+    UDP_MAX_MESSAGE_SIZE,
+)
+from raiden.settings import (
+    CACHE_TTL,
+)
+from raiden.channel.netting_channel import (
+    NODE_NETWORK_UNKNOWN,
+    NODE_NETWORK_UNREACHABLE,
+    NODE_NETWORK_REACHABLE,
+)
 from raiden.messages import decode, Ack, Ping, SignedMessage
 from raiden.utils import isaddress, sha3, pex
 
 log = slogging.get_logger(__name__)  # pylint: disable=invalid-name
 
-# - ack_result available for code that wants to block on message acknowledgment
+# - async_result available for code that wants to block on message acknowledgment
 # - receiver_address used to tie back the echohash to the receiver (mainly for
 #   logging purposes)
-WaitAck = namedtuple('WaitAck', ('ack_result', 'receiver_address'))
+WaitAck = namedtuple('WaitAck', (
+    'async_result',
+    'receiver_address',
+))
+HealthEvents = namedtuple('HealthEvents', (
+    'event_healthy',
+    'event_unhealthy',
+))
+
+# GOALS:
+# - Each netting channel must have the messages processed in-order, the
+# protocol must detect unacknowledged messages and retry them.
+# - A queue must not stall because of synchronization problems in other queues.
+# - Assuming a queue can stall, the non-healthiness of a node must not be
+# inferred from the lack of acknowledged from a single queue, but healthiness
+# may be safely inferred from it.
+# - The state of the node must be synchronized among all tasks that are
+# handling messages.
+
+
+def event_first_of(*events):
+    """ Waits until one of `events` is set.
+
+    The event returned is /not/ cleared with any of the `events`, this values
+    must not be reused if the clearing behavior is used.
+    """
+    first_finished = Event()
 
-CACHE_TTL = 60
-TTL_CACHE = cachetools.TTLCache(maxsize=50, ttl=CACHE_TTL)
+    for event in events:
+        if event is not None:
+            # ignore the underlying event
+            event.rawlink(lambda _: first_finished.set())
```

Comment:
In respect to the `None` value, not really, I'm actually going to change it to raise a ValueError instead of ignoring.